src/scf/newton.jl

```diff
@@ -0,0 +1,489 @@
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+#
+# Newton algorithm
+#
+
+"""
+    newton_step(basis::PlaneWaveBasis, φ, φproj, res, occ;
+                tol_krylov=1e-12, krylov_verbosity=1)
+
+Perform a newton step : we take as given a planewave set φ and we return the
+newton step δφ and the updated state φ - δφ
+(after proper orthonormalization) where δφ solves Jac * δφ = res
+and Jac is the Jacobian of the projected gradient descent : Jac = Ω+K.
+δφ is an element of the tangent space at φproj (set to φ if not specified in newton function).
+"""
+function newton_step(basis::PlaneWaveBasis, φ, φproj, res, occ;
+                     tol_krylov=1e-12, krylov_verbosity=1)
+
+    # necessary quantities
+    N = size(φ[1], 2)
+    Nk = length(basis.kpoints)
+
+    # compute quantites at the point which define the tangent space
+    ρproj = compute_density(basis, φproj, occ)
+    energies, Hproj = energy_hamiltonian(basis, φproj, occ; ρ=ρproj)
+
+    # packing routines
+    pack(φ) = pack_arrays(basis, φ)
+    unpack(x) = unpack_arrays(basis, x)
+    packed_proj(δx, x) = pack(proj_tangent(unpack(δx), unpack(x)))
+
+    # mapping of the linear system on the tangent space
+    function f(x)
+        δφ = unpack(x)
+        δφ = proj_tangent(δφ, φproj)
+        Kδφ = apply_K(basis, δφ, φproj, ρproj, occ)
+        Ωδφ = apply_Ω(basis, δφ, φproj, Hproj)
+        pack(Ωδφ .+ Kδφ)
+    end
+
+    # project res on the good tangent space before starting
+    res = proj_tangent(res, φproj)
+
+    # solve (Ω+K) δφ = res on the tangent space with KrylovKit
+    δφ, info = linsolve(f, pack(res);
+                        tol=tol_krylov, verbosity=krylov_verbosity,
+                        # important to specify custom orthogonaliser to keep
+                        # iterations on the tangent space
+                        orth=OrthogonalizeAndProject(packed_proj, pack(φproj)))
+    δφ = unpack(δφ)
+    # ensure proper projection onto the tangent space
+    δφ = proj_tangent(δφ, φ)
+    φ_newton = similar(φ)
+
+    # perform newton_step
+    for ik = 1:Nk
+        φ_newton[ik] = ortho_qr(φ[ik] - δφ[ik])
+    end
+    (φ_newton, δφ)
+end
+
+"""
+    newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+                    tol=1e-6, maxiter=20, φproj=nothing, verbosity=0,
+                    callback=NewtonDefaultCallback(),
+                    is_converged=NewtonConvergenceDensity(tol))
+
+Newton algorithm. Be careful that the starting needs to be not too far from the solution.
+Still contains debugging features.
+If φproj is nothing, we use the successive φ's to define the tangent spaces on which we solve
+Newton's equation. φproj can be useful when we want to fix the tangent space for testing.
+"""
+function newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+                tol=1e-6, maxiter=20, φproj=nothing, verbosity=0,
```

Comment:
There should be no verbosity argument here: the verbosity is controlled by the callback